File: js/leafxml_xform.js
"use strict";

/*
 * leafxml_xform.js
 * ================
 * 
 * JavaScript module backing leafxml_xform.html
 */

/*
 * Define XformApp namespace
 */
window.XformApp = (function() {
  
  /*
   * Local functions
   * ===============
   */
  
  /*
   * Get a document element by its ID, throwing an error if no element
   * exists with that ID.
   * 
   * Parameters:
   * 
   *   eid - string with the element ID
   * 
   * Return:
   * 
   *   the element object from the document
   */
  function getElement(eid) {
    if (typeof eid !== "string") {
      throw new Error();
    }
    
    const result = document.getElementById(eid);
    if (result === null) {
      throw new Error("Failed to find element '" + eid + "'");
    }
    
    return result;
  }
  
  /*
   * Perform a transform operation.
   * 
   * The styles are "esc0" "esc1" "esc2" "e64" "d64" which match the
   * select option values on the HTML control.
   * 
   * Parameters:
   * 
   *   input - the string to transform
   * 
   *   style - the specific transform
   * 
   * Return:
   * 
   *   the transformed text
   */
  function doTransform(input, style) {
    
    // Check parameters
    if (typeof input !== "string") {
      throw new Error();
    }
    if (typeof style !== "string") {
      throw new Error();
    }
    
    // Handle specific transform
    let result;
    if (style === "esc0") {
      // Content-text escaping
      result = LeafXML.escapeText(input, 0);
      
    } else if (style === "esc1") {
      // Single-quoted escaping
      result = LeafXML.escapeText(input, 1);
      
    } else if (style === "esc2") {
      // Double-quoted escaping
      result = LeafXML.escapeText(input, 2);
      
    } else if (style === "e64") {
      // Encode to base64
      result = LeafXML.toText64(input);
      
    } else if (style === "d64") {
      // Decode from base64
      result = LeafXML.fromText64(input);
      
    } else {
      throw new Error();
    }
    
    // Return result
    return result;
  }
  
  /*
   * Function called when a request is received to perform the transform
   * operation.
   */
  function handleTransform() {
    // Get controls
    const txtInput     = getElement("txtInput");
    const txtOutput    = getElement("txtOutput");
    const selTransform = getElement("selTransform");
    
    // Perform operation
    try {
      txtOutput.value = doTransform(txtInput.value, selTransform.value);
    } catch (ex) {
      txtOutput.value = "[Operation failed!]";
    }
  }
  
  /*
   * Public functions
   * ================
   */
  
  /*
   * Function called when the whole DOM is loaded.
   */
  function bootstrap() {
    // Attach button listeners
    const btnUpload = getElement("btnTransform");
    btnUpload.addEventListener("click", (event) => {
                                handleTransform(); });
  }
  
  /*
   * Exports
   * =======
   */
  
  return {
    bootstrap: bootstrap
  };
  
}());

/*
 * Register bootstrap function to run when document DOM loaded.
 */
window.addEventListener("DOMContentLoaded",
  (event) => { window.XformApp.bootstrap(); });

File: perl/leafxml_xform.pl
#!/usr/bin/env perl
use v5.16;
use warnings;

use LeafXML::Util qw(
  readFullText
  writeFullText
  escapeText
  toText64
  fromText64
);

=head1 NAME

leafxml_xform.pl - Text transformations using LeafXML functions.

=head1 SYNOPSIS

  leafxml_xform.pl esc 0 < input.txt > output.txt
  leafxml_xform.pl e64 < input.txt > output.txt
  leafxml_xform.pl d64 < input.txt > output.txt

=head1 DESCRIPTION

Read text from standard input, transform it using LeafXML functions, and
print the transformed text to standard output.

The input may be in UTF-8 with or without a byte order mark, or in
UTF-16 with a byte order mark.  The output is always in UTF-8 without a
byte order mark.

The C<esc> invocation performs entity escaping.  It does not, however,
verify that only valid codepoints are in use in the input.  This
invocation takes a single parameter, which must be C<0> for content text
escaping, C<1> for single-quoted attribute escaping, or C<2> for
double-quoted attribute escaping.

The C<e64> invocation encodes the text into Base64, where UTF-8 encoding
is used within the Base64 data.

The C<d64> invocation decodes text from Base64 with UTF-8 into full
UTF-8.

=cut

# ==================
# Program entrypoint
# ==================

# Get program mode
#
(scalar(@ARGV) > 0) or die "Expecting program arguments";
my $mode = shift;
(not ref($mode)) or die "Invalid argument type";

# Handle specific mode
#
if ($mode eq 'esc') {
  # Get escaping style
  (scalar(@ARGV) == 1) or die "Wrong number of arguments for esc mode";
  my $style = shift;
  (not ref($style)) or die "Invalid argument type";
  
  (($style eq '0') or ($style eq '1') or ($style eq '2')) or
    die "Unrecognized escaping style: $style";
  
  # Perform operation
  my $input;
  readFullText(\$input);
  $input = escapeText($input, int($style));
  writeFullText(\$input);
  
} elsif ($mode eq 'e64') {
  # Check arguments
  (scalar(@ARGV) == 0) or die "Not expecting arguments for e64 mode";
  
  # Perform operation
  my $input;
  readFullText(\$input);
  $input = toText64($input);
  print "$input\n";
  
} elsif ($mode eq 'd64') {
  # Check arguments
  (scalar(@ARGV) == 0) or die "Not expecting arguments for d64 mode";
  
  # Perform operation
  my $input;
  readFullText(\$input);
  $input = fromText64($input);
  writeFullText(\$input);
  
} else {
  die "Unrecognized program mode: $mode";
}
